File: htdocs/http_errors/error404.php
<?php
require_once '../../config/settings.inc.php';
require_once '../../include/myview.php';

$t = new MyView();
$t->title = "File Not Found (404)";
$t->content = <<<EOF
<h3>Requested file was not found</h3>
<img src="/images/cow404.jpg" class="img img-responsive" alt="404 Cow" />
EOF;
$t->render('single.phtml');
$ref = isset($_SERVER["HTTP_REFERER"]) ? $_SERVER["HTTP_REFERER"] : 'none';

// Since we are now running with php-fpm, we don't have access to Apache's
// errorlog, so we now send to syslog, so that we get some denoted error logged
// of 404s
openlog("iem", LOG_PID | LOG_PERROR, LOG_LOCAL1);
syslog(LOG_WARNING, "404 ". $_SERVER["REQUEST_URI"] .
		' remote: '. $_SERVER["REMOTE_ADDR"] .
		' referer: '. $ref);
closelog();
?>

File: htdocs/http_errors/error404.py
"""Our custom 404 handler."""
import sys

from pyiem.util import get_dbconn
from pyiem.templates.iem import TEMPLATE


def log_request(uri, environ):
    """Do some logging work."""
    sys.stderr.write(
        f"IEM 404 {uri} remote: {environ.get('REMOTE_ADDR')} "
        f"referer: {environ.get('HTTP_REFERER')}\n"
    )
    pgconn = get_dbconn("mesosite")
    cursor = pgconn.cursor()
    cursor.execute(
        """
        INSERT into weblog(client_addr, uri, referer, http_status)
        VALUES (%s, %s, %s, %s)
    """,
        (
            environ.get("REMOTE_ADDR"),
            uri,
            environ.get("HTTP_REFERER"),
            404,
        ),
    )
    cursor.close()
    pgconn.commit()


def application(environ, start_response):
    """mod-wsgi handler."""
    uri = environ.get("REQUEST_URI", "")
    # Special handling of ancient broken windrose behaviour
    if uri.startswith("/onsite/windrose/climate"):
        start_response("410 Gone", [("Content-type", "text/plain")])
        return [b"Resource is no longer available."]

    # We should re-assert the HTTP status code that brought us here :/
    start_response("404 Not Found", [("Content-type", "text/html")])
    content = """
<h3>Requested file was not found</h3>
<img src="/images/cow404.jpg" class="img img-responsive" alt="404 Cow" />
    """
    ctx = {"title": "File Not Found (404)", "content": content}
    try:
        log_request(uri, environ)
    except Exception as exp:
        sys.stderr.write(str(exp) + "\n")

    return [TEMPLATE.render(ctx).encode("utf-8")]
